common/onion.c

```diff
@@ -0,0 +1,305 @@
+		 *    for every node.
+		 *  - MUST include `short_channel_id` for every non-final node.
+		 */
+		tlv_amt.amt_to_forward = forward.millisatoshis; /* Raw: TLV convert */
+		tlv_cltv.outgoing_cltv_value = outgoing_cltv;
+		tlv_scid.short_channel_id = *scid;
+		tlv->amt_to_forward = &tlv_amt;
+		tlv->outgoing_cltv_value = &tlv_cltv;
+		tlv->short_channel_id = &tlv_scid;
+
+		return tlv_hop(ctx, tlv);
+	} else {
+		return v0_hop(ctx, scid, forward, outgoing_cltv);
+	}
+}
+
+u8 *onion_final_hop(const tal_t *ctx,
+		    bool use_tlv,
+		    struct amount_msat forward,
+		    u32 outgoing_cltv,
+		    struct amount_msat total_msat,
+		    const struct secret *payment_secret)
+{
+	/* These go together! */
+	if (!payment_secret)
+		assert(amount_msat_eq(total_msat, forward));
+
+	if (use_tlv) {
+		struct tlv_tlv_payload *tlv = tlv_tlv_payload_new(tmpctx);
+		struct tlv_tlv_payload_amt_to_forward tlv_amt;
+		struct tlv_tlv_payload_outgoing_cltv_value tlv_cltv;
+#if EXPERIMENTAL_FEATURES
+		struct tlv_tlv_payload_payment_data tlv_pdata;
+#endif
+
+		/* BOLT #4:
+		 *
+		 * The writer:
+		 *  - MUST include `amt_to_forward` and `outgoing_cltv_value`
+		 *    for every node.
+		 *...
+		 *  - MUST NOT include `short_channel_id` for the final node.
+		 */
+		tlv_amt.amt_to_forward = forward.millisatoshis; /* Raw: TLV convert */
+		tlv_cltv.outgoing_cltv_value = outgoing_cltv;
+		tlv->amt_to_forward = &tlv_amt;
+		tlv->outgoing_cltv_value = &tlv_cltv;
+
+#if EXPERIMENTAL_FEATURES
+		if (payment_secret) {
+			tlv_pdata.payment_secret = *payment_secret;
+			tlv_pdata.total_msat = total_msat.millisatoshis; /* Raw: TLV convert */
+			tlv->payment_data = &tlv_pdata;
+		}
+#else
+		/* Wihtout EXPERIMENTAL_FEATURES, we can't send payment_secret */
+		if (payment_secret)
+			return NULL;
+#endif
+		return tlv_hop(ctx, tlv);
+	} else {
+		static struct short_channel_id all_zero_scid;
+		/* No payment secrets in legacy format. */
+		if (payment_secret)
+			return NULL;
+		return v0_hop(ctx, &all_zero_scid, forward, outgoing_cltv);
+	}
+}
+
+/* Returns true if valid, and fills in type. */
+static bool pull_payload_length(const u8 **cursor,
+				size_t *max,
+				enum onion_payload_type *type,
+				size_t *len)
+{
+	/* *len will incorporate bytes we read from cursor */
+	const u8 *start = *cursor;
+
+	/* BOLT #4:
+	 *
+	 * The `length` field determines both the length and the format of the
+	 * `hop_payload` field; the following formats are defined:
+	 */
+	*len = fromwire_bigsize(cursor, max);
+	if (!cursor)
+		return false;
+
+	/* BOLT #4:
+	 * - Legacy `hop_data` format, identified by a single `0x00` byte for
+	 *   length. In this case the `hop_payload_length` is defined to be 32
+	 *   bytes.
+	 */
+	if (*len == 0) {
+		if (type)
+			*type = ONION_V0_PAYLOAD;
+		assert(*cursor - start == 1);
+		*len = 1 + 32;
+		return true;
+	}
+
+#if !EXPERIMENTAL_FEATURES
+	/* Only handle legacy format */
+	return false;
+#else
```

Comment:
TLV payloads support is not experimental, using it by default is IIRC.

Reply:
And payment_secret support is experimental only.

Reply:
Ok, will fix, but I thought I copied that first #if from the old code...

Reply:
Yes, I did: sphinx_parse_payload used to bail if not legacy...